internal/tmpfile/tmpfile.go

```diff
@@ -0,0 +1,66 @@
+package tmpfile
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	userOnlyRW = 0o600
+	allRWX     = 0o777
+)
+
+// KeepExtension creates a temporary file that with the same extension as the provided filename.
+// It will be created in the subdirectory "buildkite-agent" inside the system temp directory.
+func KeepExtension(filename string) (*os.File, error) {
+	return KeepExtensionWithMode("buildkite-agent", filename, userOnlyRW)
+}
+
+// KeepExtensionAndClose creates a temporary file that with the same extension as the provided
+// filename and closes it. It will be created in the subdirectory `dir` inside the system temp
+// directory.
+func KeepExtensionAndClose(dir, filename string) (string, error) {
+	return KeepExtensionWithModeAndClose(dir, filename, userOnlyRW)
+}
+
+// KeepExtensionWithMode creates a temporary file with the same extension of the
+// provided `filename` and sets its permissions to `perm`. The file will be created in the
+// subdirectory `dir` inside the system temp directory.
+func KeepExtensionWithMode(dir, filename string, perm fs.FileMode) (*os.File, error) {
+	extension := filepath.Ext(filename)
+	basename := strings.TrimSuffix(filename, extension)
+	tempDir := filepath.Join(os.TempDir(), dir)
+
+	// Best effort ensure tempDir exists
+	if _, err := os.Stat(tempDir); os.IsNotExist(err) {
```

Comment:
Why does it need to check existence of the dir beforehand? `MkdirAll` shouldn't error if it does.

Reply:
Good point, I just ported it over from the old code.

Reply:
I've done this now.